test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java

```diff
@@ -1196,6 +1198,21 @@ public static String randomDateFormatterPattern() {
         return randomFrom(FormatNames.values()).getName();
     }
 
+    /**
+     * Randomly choose between {@link EsExecutors#DIRECT_EXECUTOR_SERVICE} (which does not fork), {@link ThreadPool#generic}, and one of the
+     * other named threadpool executors.
+     */
+    public static Executor randomExecutor(ThreadPool threadPool, String... otherExecutorNames) {
```

Comment:
nit/question: is there a use of this that passes anything to `otherExecutorNames`? Having something like `genericOrDirect` for the cases that needs one of the two might make this easier to read.

Reply:
In this PR there's just one, in `JoinValidationServiceTests`. I expect to be able to adjust `org.elasticsearch.index.shard.IndexShardTests#testRunUnderPrimaryPermitDelaysToExecutorWhenBlocked` to use this too although it's a little more involved so I'll do it later.